File: summary/go/19_knapsack.go
package main

import (
	"fmt"
)

func knapsack(weights []int, values []int, capacity int) int {
	n := len(weights)
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, capacity+1)
	}

	for i := 1; i <= n; i++ {
		for w := 1; w <= capacity; w++ {
			if weights[i-1] <= w {
				dp[i][w] = max(dp[i-1][w], values[i-1]+dp[i-1][w-weights[i-1]])
			} else {
				dp[i][w] = dp[i-1][w]
			}
		}
	}
	return dp[n][capacity]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	var n, capacity int
	fmt.Scan(&n, &capacity)
	weights := make([]int, n)
	values := make([]int, n)

	for i := 0; i < n; i++ {
		fmt.Scan(&weights[i], &values[i])
	}

	result := knapsack(weights, values, capacity)
	fmt.Printf("Maximum value that can be accommodated: %d\n", result)
}

File: summary/rust/19_knapsack.rs
fn knapsack(weights: &[i32], values: &[i32], capacity: i32) -> i32 {
    let n = weights.len();
    let mut dp = vec![vec![0; (capacity + 1) as usize]; n + 1];

    for i in 1..=n {
        for w in 1..=capacity as usize {
            if weights[i - 1] as usize <= w {
                dp[i][w] = dp[i][w].max(values[i - 1] + dp[i - 1][w - weights[i - 1] as usize]);
            } else {
                dp[i][w] = dp[i - 1][w];
            }
        }
    }

    dp[n][capacity as usize]
}

fn main() {
    let weights: Vec<i32> = vec![2, 3, 4, 5];
    let values: Vec<i32> = vec![3, 4, 5, 6];
    let capacity = 5;
    let max_value = knapsack(&weights, &values, capacity);
    println!("Maximum value that can be accommodated: {}", max_value);
}
